### ISSUE TITLE
What flavour of Java?

### ISSUE THREAD

**Issue by derimmer** [vasl-developers/vasl#1402]
I have been using Java JDK 11 in my dev environment for the last number of vasl versions. This is now the minimum requirement for VASSAL and since Java JDK is now up to version 20, it feels like it is a good time to move to a higher version in our dev work. 

Does anyone more knowledgeable about Java than me, meaning all of you, have any thoughts about a preferred Java version we should move to? 

Should everyone coding VASL (pretty much @mtl1971 and @derimmer at present) be using the same version?

**Comment by uckelman** [comment]
Compiling your code using any version of Java >= 11 is fine and shouldn't give you any problems. (We compile Vassal releases with Java 19 presently.) If you compile for a target after Java 11, however, you will get a warning message from current Vassal (3.6.15) every time such a class is loaded advising you that you should recompile for Java 11 compatibility.

Presently, we bundle Java 19 with our MacOS and Windows releases. We cannot bundle Java with Linux releases due to how shared libraries work on Linux---the most reliable way for users to get Java on Linux is via their package manager. Last time I checked, there were no Java packages later than 11 for some of the still-supported Long Term Support versions of Ubuntu. Having checked just now, I see that the two oldest releases which are not yet EOL (14.04 and 16.04) have nothing later than Java 11. From 18.04 onward, there are Java 17 packages. We could advance the maximum target to Java 17 if we were willing to drop support for anything older than Ubuntu 18.04.

**Comment by derimmer** [comment]
I updated my project settings in Intelllij to java 19. Built VASL. Ran VASL from the IDE. No issues. 

I updated this part of the VASL pom.xml file from 11 to 19:

`<plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.3</version>
                <configuration>
                    <verbose>true</verbose>
                    <source>11</source>
                    <target>11</target>
                    <showDeprecation>true</showDeprecation>
                </configuration>
            </plugin>`

Then I used maven to build the VASL module. No joy. It gave me an error message about not being able to target 19. 

So, I reverted the change to the pom and ran the maven build again. This time it worked. 

But have I actually done anything? How can I tell that the maven process is compiling the release version with Java 19?

Are there other changes that I need to make to the pom.xml?

I saw @uckelman 's comment about getting warning messages. I am using VASSAL3.6.17 and have not seen any such messages so far.

**Comment by uckelman** [comment]
> I updated this part of the VASL pom.xml file from 11 to 19:

That is exactly the part you should _not_ change from 11 to 19. The `source` and `target` properties are for compatibility with Java 11. They do not select which compiler you're using, only the minimum Java version the compiled class files should target.

> I saw @uckelman 's comment about getting warning messages.

You probably didn't succeed at compiling with `target` set to 19, in that case.

`./mvnw -v` will make Maven print some version information, including the version of Java it's using.

I see this:
```
[uckelman@scylla trunk]$ ./mvnw -v 
Warning: JAVA_HOME environment variable is not set.
Apache Maven 3.8.5 (3599d3414f046de2324203b78ddcf9b5e4388aa0)
Maven home: /home/uckelman/.m2/wrapper/dists/apache-maven-3.8.5-bin/67203e94/apache-maven-3.8.5
Java version: 20, vendor: Red Hat, Inc., runtime: /usr/lib/jvm/java-20-openjdk-20.0.0.0.36-1.rolling.fc37.x86_64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "6.2.12-200.fc37.x86_64", arch: "amd64", family: "unix"
```

(Incidentally we're using 20 now for building Vassal releases, from 3.6.17 onward.)

**Comment by derimmer** [comment]
When I run mvnw -v, I get told that I am using Java version 11.0.12, which is the version I have been using. So @uckelman  you are correct that I did not succeed in compiling with target set to 19. 

Any thoughts as to where and when I should make that targeting explicit? Not the pom apparently. Do I need to edit mvnw?

**Comment by uckelman** [comment]
The Java that Maven uses is the one specified in the `JAVA_HOME` environment variable, and if that's not set, it will use whatever the Java tools are first on the `PATH`.

I don't have `JAVA_HOME` set. On my system (Fedora Linux), `/usr/bin/javac` is a symlink to the place where Java 20 is installed, and the symlink is controlled by `alternatives`, which lets you choose which of several installed versions the symlink will point to. If I wanted to change permanently which version of Java Maven would use, I'd run `alternatives` to do that.

I don't know what kind of system you're using. If you don't need any other versions of Java, the simplest thing might be to uninstall all the ones you don't intend for Maven to use.

**Comment by derimmer** [comment]
Since I had JAVA_HOME set, I changed it to 19. Mvnw -v then displayed 19.0.2 as the Java version. Ran the maven build with success. I got one warning message:

[WARNING] system modules path not set in conjunction with -source 11

I found a couple of other places in my IDE that I had forgotten to change from Java 11 to 19, notably in the Run Configurations. 

Now when I compile the VASL module from within the IDE I see messages that Java 19 is being used. Same when I run it from within the IDE. 

The version built with maven runs from the VASSAL window (3.6.17).

All seems fine now. However, I am still not getting the warning messages about recompiling for Java 11 compatability.

**Comment by derimmer** [comment]
@uckelman thanks for your continued and excellent help on this. Much appreciated!

**Comment by uckelman** [comment]
> However, I am still not getting the warning messages about recompiling for Java 11 compatability.

You won't, so long as you leave the `source` and `target` set in `pom.xml` set to `11`.

**Comment by uckelman** [comment]
```
[WARNING] system modules path not set in conjunction with -source 11
```
That warning is a result of not setting `-bootclasspath` or `--release`. We should probably do that, but I haven't had time to look into it.

**Comment by uckelman** [comment]
I've modified Vassal's builds to use `--release` instead of `--source` and `--target`: https://github.com/vassalengine/vassal/pull/12272/files

The explanation for the difference is here: https://stackoverflow.com/a/43103038/181106

Here's a PR for you which does the same:

https://github.com/vasl-developers/vasl/pull/1413

**Comment by derimmer** [comment]
I have merged the PR. Seems like we are on 19! 

I will sit here for a bit and see if 19 causes any issues and will move to 20 (or wherever VASSAL is) when I build the release version of 6.6.7 in the Fall.

**Comment by derimmer** [comment]
Too good to be true. The IDE shows an error for the revised POM and when I tried to do a maven build with it, maven fails. 

For once, I was able to puzzle it out myself. 

I needed to update the apache-maven-compiler version in the POM. Then I had to add the new version to the project in IntelliJ. 

Now we are cooking with gas. I will push a commit with the revised POM.

We are at 19. Time to close this.
